osu.Game/Rulesets/UI/DrawableRulesetDependencies.cs

```diff
@@ -190,24 +196,27 @@ protected override void Dispose(bool disposing)
             }
         }
 
-        private class FallbackShaderManager : ShaderManager
+        private class RulesetShaderManager : ShaderManager
         {
-            private readonly ShaderManager primary;
-            private readonly ShaderManager fallback;
+            private readonly ShaderManager parent;
 
-            public FallbackShaderManager(IRenderer renderer, ShaderManager primary, ShaderManager fallback)
-                : base(renderer, new ResourceStore<byte[]>())
+            public RulesetShaderManager(IRenderer renderer, NamespacedResourceStore<byte[]> rulesetResources, ShaderManager parent)
+                : base(renderer, rulesetResources)
             {
-                this.primary = primary;
-                this.fallback = fallback;
+                this.parent = parent;
             }
 
-            public override byte[]? LoadRaw(string name) => primary.LoadRaw(name) ?? fallback.LoadRaw(name);
-
-            protected override void Dispose(bool disposing)
+            public override IShader Load(string vertex, string fragment)
             {
-                base.Dispose(disposing);
-                if (primary.IsNotNull()) primary.Dispose();
+                try
+                {
+                    return base.Load(vertex, fragment);
+                }
+                catch
+                {
+                    // Shader lookup is very non-standard. Rather than returning null on missing shaders, exceptions are thrown.
+                    return parent.Load(vertex, fragment);
+                }
```

Comment:
This code as written currently crashes for me when trying to start an osu! map. It does so because:

1. Rulesets will attempt to load framework shaders (because they can't see the game-global shader manager, they see the ruleset shader manager)
2. Upon doing so, they will hit the `base.Load()` call here
3. `base.Load()` internally calls `LoadRaw()`
4. `LoadRaw()` can't read framework resources, so it helpfully returns null source
5.
```
System.ArgumentNullException: Value cannot be null. (Parameter 'source')
   at Vortice.D3DCompiler.Compiler.Compile(ReadOnlySpan`1 source, ShaderMacro[] defines, Include include, String entryPoint, String sourceName, String profile, ShaderFlags shaderFlags, Blob& blob, Blob& errorBlob)
   at Veldrid.D3D11.D3D11Shader.CompileCode(ShaderDescription description)
   at Veldrid.D3D11.D3D11Shader..ctor(ID3D11Device device, ShaderDescription description)
   at Veldrid.D3D11.D3D11ResourceFactory.CreateShaderCore(ShaderDescription& description)
   at Veldrid.ResourceFactory.CreateShader(ShaderDescription& description)
   at Veldrid.ResourceFactory.CreateShader(ShaderDescription description)
   at osu.Framework.Graphics.Veldrid.Shaders.VeldridShader.loadToGpu() in D:\Open Source\osu-framework\osu.Framework\Graphics\Veldrid\Shaders\VeldridShader.cs:line 214
```

This can be avoided by also overriding `LoadRaw()` to try the parent, too - but, more importantly, wasn't the goal here to *avoid* recompiling the game-global shaders? As is they will get recompiled every play anyway because every play is a fresh instance of `RulesetShaderManager`, and the base call comes first. What am I missing here?

Reply:
I'm quite confused. How are you seeing the crash? Are you potentially running in debug with "throw on caught exceptions" turned on or something like that?

There should be a try-catch which falls back to the parent (global) shader manager. Each `ShaderManager` has its own cache, so when falling back, there would be no recompilation as it is cached at that level.

As such, overriding `LoadRaw` should not be required.

Reply:
I just did a framework checkout and ran the game in debug. I didn't touch any other settings. Is that an invalid way to test?

Reply:
Are you able to provide a full call stack of the crash? Is it somehow happening outside the `try-catch` in this PR?

I can't repro any crash locally, unfortunately..

Reply:
Here's a full log: [runtime.log](https://github.com/ppy/osu/files/11096663/runtime.log)

Reply:
Very interesting. That might be D3D specific, but it does reveal that we are still attempting to load a shader to the GPU even after it fails to compile, which is a large oversight.

https://github.com/ppy/osu-framework/blob/e84265d661847a21fbf27875f3528f35f209bf2d/osu.Framework/Graphics/Veldrid/Shaders/VeldridShader.cs#L202-L205

This probably needs to throw rather than gracefully being caught and continuing with execution.

Or at very least, the scheduled gpu upload should be omitted in fail cases.

Reply:
I've made adjustments to the framework-side PR to address the issue you found. Turns out I can reproduce it with metal renderer, and was using legacy GL when testing this PR.

Reply:
Tested with latest framework changes, looks good now 👍